### ISSUE TITLE
Slow persistent pickle with large persistent._seen_ever

### ISSUE THREAD

**Issue by Andykl** [renpy/renpy#3042]
Our game has 310_000 nodes in `renpy.game.script.namemap` and if I pass the whole game, the size of `persistent._seen_...` (of all sorts) becomes so large that it takes over 5 seconds on my machine to perform `renpy.save_persistent`, whereas with these sets empty it's only 0.1-0.2 seconds.

My suggestion is to put all the RenPy defined persistent data into a separate class (RenpyPersistent or another name) and instead of pickle this data, use marshal. This is possible because only the built-in types are used there. Using marshal even on fully populated sets takes on average 0.2 seconds on my machine.

The format for saving data has 2 options

1. Leave a single file. Approximate new `persistent` file format
```
RENPY PERSISTENT (header)
struct.pack("I", len(Pickled Persistent object))
Pickled Persistent object
Maybe some separator
Marshaled RenpyPersistent data
```

2. Add a new file (renpy_persistent or some other name)

I personally prefer the second option, because it's easier to implement, and as a sideeffect it allows developers to distribute a "100% persistent file" which will not break user settings.

@renpytom if you generally approve the idea, I may come with implementation shortly.

**Comment by Gouvernathor** [comment]
The current persistent system allows for classes defined in `python early` to be used, see [this](https://www.renpy.org/doc/html/persistent.html#persistent-data). I believe Marshal wouldn't allow this, which is a big restriction.
Do you plan to toggle the use of marshal, via a config variable like `config.use_cpickle` for example ?
If not, a hybrid version could still be imagined, where the persistent is tried to be marshaled, and if it fails because of types, the exception is caught and reverts back to pickle.

Also there is the caution given in [its doc](https://docs.python.org/3/library/marshal.html) that (if I understand it correctly) data saved in one version of python may not be recoverable in a future version of python. Given that save files are supposed to be long-term compatible, and given that we're supposed to move to PY3 by a few months from now, I think this is worth worrying about.

**Comment by uyjulian** [comment]
I wonder if it would be possible to marshal on main thread, then pickle on separate thread?

**Comment by renpytom** [comment]
@Andykl , 

Do you have experiments that show that marshal is fast enough to make a difference? I put together a test program, 

https://gist.github.com/renpytom/c6edae881c843c99bd8bb99617ce6e98

that creates a dictionary meant to mimic seen_labels, and then pickles/marshalls it. On my machine, a dictionary with 1,000,000 entries takes about 366ms to pickle, while marshalling it takes 262ms. (That's just for the pickling/marshalling, not the compressed save to disk.) 

I'm not fundamentally opposed to the idea of renpy_persistent, but at the same time I'm not completely convinced that marshalling will provide enough of a performance improvement to make it worth doing. I'd like to see some sort of experiments that show the change actually will speed things up.

Basically, I see the problem, and while I'd like to be sure it persists in 7.4.9, I think it's worth solving. I'm not sure something like marshal is the right way to do this, versus something more exotic. (Like writing out the changes on each run.) 

@Gouvernathor Andy discussed  Ren'Py defined persistent data (persistent._seen labels, etc), so I don't think it's necessary to talk about marshal's limitation. I'm not too worries about marshal changing format - we could always re-implement it in cython if it did.

@uyjulian We already do that on autosave. I don't think it makes a lot of sense to do that on real saves - a person might follow real saving by sleeping their phone or computer, and for persistent it would be exiting Ren'Py. In both cases, we'd want the time the save is done to be clear.

**Comment by Booplicate** [comment]
Please, if possible, don't split saves into 2 files. It'd be highly inconvenient for people who want to transfer saves. And it makes everything messier, too.

**Comment by renpytom** [comment]
@Booplicate This is for persistent data, not saves.

**Comment by multimokia** [comment]
I don't think that's quite what @Booplicate means. If the user wished to transfer their progress from 1 PC to another, they'll also need their persistents. Transferring becomes more of a hassle for them as a result of that.

**Comment by uyjulian** [comment]
There is no split of data

**Comment by renpytom** [comment]
We will probably be splitting it into two files - that's the point of this issue. 

And I don't think that transferring two files is that much harder than transferring one, especially because the two files can be transferred independently.

**Comment by Booplicate** [comment]
For the end-users, it's a big issue.

**Comment by uyjulian** [comment]
You have 2 files.

Click once on 1st file, shift click on 2nd file, drag and drop to desired destination.

If user have issue with transfering 1 file, they will have issue with 2 file, so it may be better to create utility for transfer file to other location to make user easy.

**Comment by renpytom** [comment]
Can you show an example of a user that is able to transfer one file, but not two? I can't think of any case where that would happen. 

(I'm also not that interested in designing Ren'Py around what computer-illterate people are capable of.)

**Comment by Booplicate** [comment]
I am much aware of how to transfer files, but there are people who aren't. If they are struggling now, it's not a reason to make it worse.

**Comment by uyjulian** [comment]
If the transfer file issue is something necessary you can create a utility to transfer file so you can reduce the support request needed.

**Comment by Booplicate** [comment]
> Can you show an example of a user that is able to transfer one file, but not two?

I could invite you to our TS if you want. I'm quite serious. People just want to play the game, not think of which of 2 files (or both?) they need to move. Yes, even adding 1 file to the equation makes everything worse.

> (I'm also not that interested in designing Ren'Py around what computer-illterate people are capable of.)

I understand, but the point of this is to be convenient for the end-users as much as possible. Optimization is important, but it shouldn't go in a way of users' comfort.

> If the transfer file issue is something necessary you can create a utility to transfer file so you can reduce the support request needed.

Doesn't really sound relevant. How do you imagine a tool that transfers game saves from 1 PC to another. We're developing games, not FTP clients.

**Comment by multimokia** [comment]
I also want to add, if the file copy doesn't go properly i.e. files get mismatched (one new persistent, one old one) you now have data integrity issues to manage which leads to another whole host of bugs as a result of it, which are also harder to trace back

**Comment by renpytom** [comment]
@Booplicate I don't know what a TS is, but they could just put the game on a dropbox (or something similar) and everything would work. I don't believe that for a minimally computer literate person, copying two files is more of a burden than copying one.

@multimokia That's a legitimate concern, though I'd say it's fine if the game keeps working when it's mismatched. (Which the current design supports.) 

I need to look at this a bit more - it might be better to append the marshall to the persistent file - but I 100% don't want to make "what will a computer-illterate user do?" be the deciding factor.

**Comment by multimokia** [comment]
> though I'd say it's fine if the game keeps working when it's mismatched

Would this be configurable? I see cases where this leads to continuity issues in projects which lead to bigger issues in terms of flows/cohesion and immersion. I'd rather this not be allowed to work really.

**Comment by Booplicate** [comment]
That said, if this is unavoidable change, can it at least be toggled via a config variable? So we don't have to rewrite the entire system to keep the data safe.

**Comment by mal** [comment]
Is it worth investigating the impact of improvements to `pickle` that will be available in python3; given that it's the next big dev push and may impact this issue sufficiently to potentially avoid needing to solve it by complicating and potentially limiting compatibility of persistent saves files?

There is also potential to gain far more significant improvement by altering the data structure prior to serialisation. For example flattening the keys into a single tuple (`tuple(v for k in data for v in k)`) prior pickling yields excellent results in py3 (anecdotally the same data volume in ~1/20th the time, and quicker in py3 pickle than py3 marshal), for a small complexity trade off during load to reassemble the structure: `a = pickle.loads(tmp); dict.fromkeys(zip(a[::3], a[1::3], a[2::3]), True)`. This is by no means comprehensive, but it seems at least worth looking further afield when tackling this than only at the serialisation lib.

**tl;dr:** There could be significant benefit to looking further afield than only the serialisation library when it comes to tackling this problem, specifically at the structures being saved.

**Comment by Andykl** [comment]
Pickle in Python 3 works way faster, so it can be closed.
